![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)


# Proyecto: Riesgo en el Banco Giturra

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### Cuerpo Docente:

- Profesor: Gabriel Iturra, Ignacio Meza De La Jara
- Auxiliar: Sebastián Tinoco
- Ayudante: Arturo Lazcano, Angelo Muñoz

### Estudiante:
- Mario González Otero

### Link de repositorio de GitHub:
- [ENLACE](https://github.com/mgzotero/MDS7202)

_Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir._

---

## Reglas

- Fecha de entrega: 19/12/2023
- **Grupos de 2 personas.**
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Estrictamente prohibida la copia.
- Pueden usar cualquier material del curso que estimen conveniente.


---


# Presentación del Problema


![](https://www.diarioeldia.cl/u/fotografias/fotosnoticias/2019/11/8/67218.jpg)


**Giturra**, un banquero astuto y ambicioso, estableció su propio banco con el objetivo de obtener enormes ganancias. Sin embargo, su reputación se vio empañada debido a las tasas de interés usureras que imponía a sus clientes. A medida que su banco crecía, Giturra enfrentaba una creciente cantidad de préstamos impagados, lo que amenazaba su negocio y su prestigio.

Para abordar este desafío, Giturra reconoció la necesidad de reducir los riesgos de préstamo y mejorar la calidad de los préstamos otorgados. Decidió aprovechar la ciencia de datos y el análisis de riesgo crediticio. Contrató a un equipo de expertos para desarrollar un modelo predictivo de riesgo crediticio.

Cabe señalar que lo modelos solicitados por el banquero deben ser interpretables. Ya que estos le permitira al equipo comprender y explicar cómo se toman las decisiones crediticias. Utilizando visualizaciones claras y explicaciones detalladas, pudieron identificar las características más relevantes, le permitirá analizar la distribución de la importancia de las variables y evaluar si los modelos son coherentes con el negocio.

Para esto Giturra les solicita crear un modelo de riesgo disponibilizandoles una amplia gama de variables de sus usuarios: como historiales de crédito, ingresos y otros factores financieros relevantes, para evaluar la probabilidad de incumplimiento de pago de los clientes. Con esta información, Giturra podra tomar decisiones más informadas en cuanto a los préstamos, ofreciendo condiciones más favorables a aquellos con menor riesgo de impago.


## Instalación de Librerías y Carga de Datos.


Para el desarrollo de su proyecto, utilice el conjunto de datos `dataset.pq` para entrenar un modelo de su elección. Se le recomienda levantar un ambiente de `conda` para instalar las librerías y así evitar cualquier problema con las versiones.


---

## Secciones Requeridas en el Informe

La siguiente lista detalla las secciones que debe contener su notebook para resolver el proyecto. 
Es importante que al momento de desarrollar cada una de las secciones, estas sean escritas en un formato tipo **informe**, donde describan detalladamente cada uno de los puntos realizados.

### 1. Introducción [0.5 puntos]

_Esta sección es literalmente una muy breve introducción con todo lo necesario para entender que hicieron en su proyecto._

- Describir brevemente el problema planteado (¿Qué se intenta predecir?)
- Describir brevemente los datos de entrada que les provee el problema.
- Describir las métricas que utilizarán para evaluar los modelos generados. Eligan **una métrica** adecuada para el desarrollo del proyecto **según la tarea que deben resolver y la institución a la cuál será su contraparte** y luego justifiquen su elección. Considerando que los datos presentan desbalanceo y que el uso de la métrica 'accuracy' sería incorrecto, enfoquen su elección en una de las métricas precision, recall o f1-score y en la clase que será evaluada.
- [Escribir al final] Describir brevemente el modelo que usaron para resolver el problema (incluyendo las transformaciones intermedias de datos).
- [Escribir al final] Indicar si lograron resolver el problema a través de su modelo. Indiquen además si creen que los resultados de su mejor modelo son aceptables y como les fue con respecto al resto de los equipos.



### 2. Carga de datos Análisis Exploratorio de Datos [Sin puntaje]

_La idea de esta sección es que cargen y exploren el dataset para así obtener una idea de como son los datos y como se relacionan con el problema._

Cargue los datos y realice un análisis exploratorio de datos para investigar patrones, tendencias y relaciones en un conjunto de datos. Se adjuntan diversos scripts para abodar rápidamente este punto. La descripción de las columnas las pueden encontrar en el siguiente [enlace](https://www.kaggle.com/datasets/parisrohan/credit-score-classification).

**NO deben escribir nada**, solo ejecutar el código y encontrar los patrones con los cuales se basaran para generar el modelo.


In [8]:
# Librerías

import pandas as pd

SEED = 1245

In [3]:
# Lectura de datos

df = pd.read_parquet(
    "dataset.pq"
)

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12500 entries, 0 to 12499
Data columns (total 22 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   customer_id               12500 non-null  object 
 1   age                       12500 non-null  float64
 2   occupation                12500 non-null  object 
 3   annual_income             12500 non-null  float64
 4   monthly_inhand_salary     10584 non-null  float64
 5   num_bank_accounts         12500 non-null  int64  
 6   num_credit_card           12500 non-null  int64  
 7   interest_rate             12500 non-null  int64  
 8   num_of_loan               12500 non-null  float64
 9   delay_from_due_date       12500 non-null  int64  
 10  num_of_delayed_payment    11660 non-null  float64
 11  changed_credit_limit      12246 non-null  float64
 12  num_credit_inquiries      12243 non-null  float64
 13  outstanding_debt          12500 non-null  float64
 14  credit

In [9]:
df.head()

customer_id   age    occupation  annual_income  monthly_inhand_salary  \
0   CUS_0xd40  23.0     Scientist       19114.12            1824.843333   
1  CUS_0x21b1  28.0       Teacher       34847.84            3037.986667   
2  CUS_0x2dbc  34.0      Engineer      143162.64           12187.220000   
3  CUS_0xb891  55.0  Entrepreneur       30689.89            2612.490833   
4  CUS_0x1cdb  21.0     Developer       35547.71            2853.309167   

   num_bank_accounts  num_credit_card  interest_rate  num_of_loan  \
0                  3                4              3          4.0   
1                  2                4              6          1.0   
2                  1                5              8          3.0   
3                  2                5              4       -100.0   
4                  7                5              5       -100.0   

   delay_from_due_date  ...  num_credit_inquiries  outstanding_debt  \
0                    3  ...                   4.0            809.98   
1                    3  ...                   2.0            605.03   
2                    8  ...                   3.0           1303.01   
3                    4  ...                   4.0            632.46   
4                    1  ...                   4.0            943.86   

   credit_utilization_ratio  credit_history_age  payment_of_min_amount  \
0                 23.933795                 NaN                     No   
1                 32.933856                27.0                     No   
2                 38.374753                18.0                     No   
3                 27.332515                17.0                     No   
4                 25.862922                31.0                    Yes   

   total_emi_per_month amount_invested_monthly  \
0            49.574949               24.785217   
1            18.816215              218.904344   
2           246.992319            10000.000000   
3            16.415452              125.617251   
4             0.000000              181.330901   

                  payment_behaviour  monthly_balance credit_score  
0  High_spent_Medium_value_payments       358.124168            0  
1    Low_spent_Small_value_payments       356.078109            0  
2   High_spent_Small_value_payments       895.494583            0  
3   High_spent_Small_value_payments       379.216381            0  
4   High_spent_Small_value_payments       364.000016            0  

[5 rows x 22 columns]

In [11]:
df.sample(5, random_state=SEED)

customer_id   age  occupation  annual_income  monthly_inhand_salary  \
1876   CUS_0xb0fd  31.0   Architect       10250.20             873.183333   
11345  CUS_0x4282  23.0  Accountant       32507.76            2918.980000   
2320   CUS_0x4db9  19.0  Accountant       53587.86            4197.655000   
121    CUS_0x8a04  37.0  Journalist       18717.02                    NaN   
7270   CUS_0x43a7  32.0    Engineer       32151.76            2591.313333   

       num_bank_accounts  num_credit_card  interest_rate  num_of_loan  \
1876                   7                4             19          4.0   
11345                  6                3             10          3.0   
2320                   6                6             18          3.0   
121                    8                5             15          5.0   
7270                   7               10             29          9.0   

       delay_from_due_date  ...  num_credit_inquiries  outstanding_debt  \
1876                    21  ...                   7.0           1437.24   
11345                   18  ...                   4.0            384.04   
2320                    28  ...                   6.0            636.73   
121                      4  ...                  10.0           2103.99   
7270                    38  ...                  15.0           2598.41   

       credit_utilization_ratio  credit_history_age  payment_of_min_amount  \
1876                  38.667698                 8.0                     NM   
11345                 30.839838                19.0                    Yes   
2320                  30.792261                15.0                    Yes   
121                   29.050667                15.0                    Yes   
7270                  21.923974                 7.0                     NM   

       total_emi_per_month amount_invested_monthly  \
1876             30.957340                     NaN   
11345         40136.000000               74.400913   
2320            124.915143              228.520514   
121              60.646993               66.720449   
7270            181.965073              268.045565   

                     payment_behaviour  monthly_balance credit_score  
1876    Low_spent_Large_value_payments       298.281651            1  
11345   Low_spent_Large_value_payments       434.372958            0  
2320   High_spent_Small_value_payments       326.329843            0  
121     Low_spent_Small_value_payments       292.107724            0  
7270    Low_spent_Small_value_payments        99.120695            1  

[5 rows x 22 columns]

In [16]:
df.select_dtypes(include="number").describe().T

count          mean           std           min  \
age                       12500.0  1.057718e+02  6.645027e+02 -5.000000e+02   
annual_income             12500.0  1.616206e+05  1.297842e+06  7.005930e+03   
monthly_inhand_salary     10584.0  4.186635e+03  3.173690e+03  3.036454e+02   
num_bank_accounts         12500.0  1.693992e+01  1.143508e+02 -1.000000e+00   
num_credit_card           12500.0  2.317272e+01  1.320059e+02  0.000000e+00   
interest_rate             12500.0  7.321336e+01  4.686822e+02  1.000000e+00   
num_of_loan               12500.0  3.099440e+00  6.510528e+01 -1.000000e+02   
delay_from_due_date       12500.0  2.106088e+01  1.486309e+01 -5.000000e+00   
num_of_delayed_payment    11660.0  3.293542e+01  2.374377e+02 -3.000000e+00   
changed_credit_limit      12246.0  1.039858e+01  6.799253e+00 -6.490000e+00   
num_credit_inquiries      12243.0  2.629233e+01  1.818210e+02  0.000000e+00   
outstanding_debt          12500.0  1.426220e+03  1.155169e+03  2.300000e-01   
credit_utilization_ratio  12500.0  3.234926e+01  5.156815e+00  2.010077e+01   
credit_history_age        11380.0  1.823040e+01  8.302078e+00  0.000000e+00   
total_emi_per_month       12500.0  1.488394e+03  8.561450e+03  0.000000e+00   
amount_invested_monthly   11914.0  6.387987e+02  2.049195e+03  0.000000e+00   
monthly_balance           12145.0 -2.744614e+22  3.024684e+24 -3.333333e+26   
credit_score              12500.0  2.881600e-01  4.529241e-01  0.000000e+00   

                                   25%           50%           75%  \
age                          25.000000     33.000000     42.000000   
annual_income             19453.327500  37572.375000  72690.215000   
monthly_inhand_salary      1622.408646   3087.595000   5967.937500   
num_bank_accounts             3.000000      6.000000      7.000000   
num_credit_card               4.000000      5.000000      7.000000   
interest_rate                 8.000000     14.000000     20.000000   
num_of_loan                   1.000000      3.000000      5.000000   
delay_from_due_date          10.000000     18.000000     28.000000   
num_of_delayed_payment        9.000000     14.000000     18.000000   
changed_credit_limit          5.370000      9.410000     14.940000   
num_credit_inquiries          4.000000      6.000000     10.000000   
outstanding_debt            566.072500   1166.155000   1945.962500   
credit_utilization_ratio     28.066517     32.418953     36.623650   
credit_history_age           12.000000     18.000000     25.000000   
total_emi_per_month          31.496968     72.887628    169.634826   
amount_invested_monthly      73.736810    134.093193    261.664256   
monthly_balance             270.150125    339.388524    471.424536   
credit_score                  0.000000      0.000000      1.000000   

                                   max  
age                       8.678000e+03  
annual_income             2.383470e+07  
monthly_inhand_salary     1.520463e+04  
num_bank_accounts         1.756000e+03  
num_credit_card           1.499000e+03  
interest_rate             5.789000e+03  
num_of_loan               1.495000e+03  
delay_from_due_date       6.700000e+01  
num_of_delayed_payment    4.293000e+03  
changed_credit_limit      3.697000e+01  
num_credit_inquiries      2.554000e+03  
outstanding_debt          4.998070e+03  
credit_utilization_ratio  4.819982e+01  
credit_history_age        3.300000e+01  
total_emi_per_month       8.197100e+04  
amount_invested_monthly   1.000000e+04  
monthly_balance           1.463792e+03  
credit_score              1.000000e+00

In [15]:
df.select_dtypes(include="object").describe().T

count unique                             top  freq
customer_id            12500  12500                       CUS_0xd40     1
occupation             12500     16                         _______   880
payment_of_min_amount  12500      3                             Yes  6571
payment_behaviour      12500      7  Low_spent_Small_value_payments  3202


### 3. Preparación de Datos [0.5 puntos]

_Esta sección consiste en generar los distintos pasos para preparar sus datos con el fin de luego poder crear su modelo._

#### 3.1 Preprocesamiento con `ColumnTransformer`

- Convierta las columnas mal leidas a sus tipos correspondientes (float, str, etc...)
- Genere un `ColumnTransformer` que:
  - Preprocese datos categóricos y ordinales.
  - Escale/estandarice datos numéricos.
  - Uitlice `.set_output(transform="pandas")` sobre su `ColumnTransformer` para setear el formato de salida a de las transformaciones a pandas.

- Luego, pruebe las transformaciones utilizando `fit_transform`.

- Posteriormente, ejecute un Holdout que le permita más adelante evaluar los modelos.

#### 3.2 Holdout 

Ejecute `train_test_split` para generar un conjunto de entrenamiento, validacióny de prueba. 

#### 3.3 Datos nulos.

Como habrá visto, existe la posibilidad de que algunos datos sean nulos. En esta sección se le solicita justificar, previo a comenzar el modelado, decidir si conservar e imputar los datos nulos o eliminar las filas. 

Note que la decisión que tomen aquí puede afectar fuertemente el rendimiento de los modelos. 
Y como siempre, más adelante tienen el espacio para experimentar con ambas opciones.

#### 3.4 Feature Engineering [Bonus - 0.5 puntos]

En esta sección, se espera que apliquen su conocimiento y creatividad para identificar y construir características que brinden una mejor orientación a su modelo para identificar los casos deseados. Para motivar la construcción de nuevas características, se recomienda explorar las siguientes posibilidades:

- Generar ratios que relacionen variables categóricas con numéricas. Estos ratios permiten capturar relaciones proporcionales o comparativas entre diferentes categorías y valores numéricos.
- Combinación de rankings entre variables numéricas y categóricas.
- Discretización de variables numéricas a categóricas.
- Etc...

**Importantes**: Al explorar estas posibilidades no se limiten solo a estas propuestas, pueden aplicar otras técnicas de feature engineering pertinentes para mejorar la capacidad de su modelo para comprender y aprovechar los patrones presentes en los datos. 



### 4. Baseline [1.5 puntos]

_En esta sección deben crear los modelos más básicos posibles que resuelvan el problema dado. La idea de estos modelos son usarlos como comparación para que en el siguiente paso lo puedan mejorar._

Implemente, entrene y evalúe varias `Pipeline` enfocadas en resolver el problema de clasificación en donde la diferencia entre estas sea el modelo utilizado.


Para esto, cada Pipeline debe:

- Tener el `ColumnTransformer` implementado en la sección anterior como primer paso.
- Implementar un imputador en caso de haber decidido conservar los datos nulos.
- Implementar un clasificador en la salida (ver siguiente lista).
  
Y además, 
- Ser evaluado de forma general imprimiendo un `classification_report`.
- Calcular y guardar la métrica seleccionada en el punto 1.2 en un arreglo de métricas (guardar nombre y valor de la métrica).

Lo anterior debe ser implementado utilizando los siguientes modelos:

- `Dummy` con estrategia estratificada.
- `LogisticRegression`.
- `KNeighborsClassifier`.
- `DecisionTreeClassifier`
- `SVC`
- `RandomForestClassifier` 
- `LightGBMClassifier` (del paquete `lightgbm`)
- `XGBClassifier` (del paquete `xgboost`).


Luego, transformando el diccionario de las métricas a un pandas `DataFrame`, ordene según los valores de su métrica de mayor a menor y responda.
- ¿Hay algún clasificador entrenado mejor que el azar (`Dummy`)?
- ¿Cuál es el mejor clasificador entrenado?
- ¿Por qué el mejor clasificador es mejor que los otros?
- Respecto al tiempo de entrenamiento, con cual cree que sería mejor experimentar (piense en el tiempo que le tomaría pasar el modelo por una grilla de optimización de hiperparámetros).

**Nota**: Puede utilizar un for más una lista con las clases de los modelos mencionados para simplificar el proceso anterior.




### 5. Optimización del Modelo [1.5 puntos]

_En esta sección deben mejorar del modelo de clasificación al variar los algoritmos/hiperparámetros que están ocupando._

- Instanciar dos nuevas `Pipeline`, similares a la anterior, pero ahora enfocada en buscar el mejor modelo. Para esto, la pipelines debe utilizar el primer y segundo mejor modelo encontrado en el paso anterior.
- Usar **`Optuna`** para tunear hiperparámetros
- **Importante**: Recuerden setear la búsqueda para optimizar la métrica seleccionada en los puntos anteriores.

Algunas ideas para mejorar el rendimiento de sus modelos:

- Agregar técnicas de seleccion de atributos/características. El parámetro de cuántas características se seleccionan debe ser parametrizable y configurado por el optimizador de hiperparámetros.
- Variar el imputador de datos en caso de usarlo.

#### Bonus

1. **Visualización con Optuna** [0.2 extras]: Explore la documentación de visualización de Optuna en el siguiente [link](https://optuna.readthedocs.io/en/stable/reference/visualization/index.html) y realice un análisis sobre el proceso de optimización de hiperparámetros realizado.
2. **Imabalanced learn** [0.3 extras]: Al ser el problema desbalanceado, pueden probar técnicas para balancear automáticamente el dataset previo a ejecutar el modelo. Para esto, puede probar con los mecanismos implementados en la librería [Imbalanced learn](https://imbalanced-learn.org/). 
3. **Probar pycaret (AutoML)** [0.3 extras].

Algunas notas interesantes sobre este proceso:

- No se les pide rendimientos cercanos al 100% de la métrica para concretar exitosamente el proyecto. Por otra parte, celebren cada progreso que obtengan.
- **Hacer grillas computables**: Si la grilla se va a demorar 1/3 la edad del universo en explorarse completamente, entonces achíquenla a algo que sepan que va a terminar.
- Aprovechen el procesamiento paralelo (con `njobs`) para acelerar la búsqueda. Sin embargo, si tienen problemas con la memoria RAM, reduzca la cantidad de jobs a algo que su computador/interprete web pueda procesar.

**Al final de este proceso, seleccione el mejor modelo encontrado, prediga el conjunto de prueba y reporte sus resultados.**

### 6. Interpretabilidad [1.0 puntos]

_En esta sección, se espera que los estudiantes demuestren su capacidad para explicar cómo sus modelos toman decisiones utilizando los datos. Dentro del análisis de interpretabilidad propuesto para el modelo, deberán ser capaces de:_

- Proponer un método para analizar la interpretabilidad del modelo. Es crucial que puedan justificar por qué el método propuesto es el más adecuado y explicar los alcances que podría tener en su aplicación.
- Identificar las características más relevantes del modelo. ¿La distribución de importancia es coherente y equitativa entre todas las variables?
- Analizar 10 observaciones aleatorias utilizando un método específico para verificar la coherencia de las interacciones entre las características.
- Explorar cómo se relacionan las variables utilizando algún descriptivo de interpretabilidad.
- ¿Existen variables irrelevantes en el problema modelado?, ¿Cuales son?.

Es fundamental que los estudiantes sean capaces de determinar si su modelo toma decisiones coherentes y evaluar el impacto que podría tener la aplicación de un modelo con esas variables en una población. ¿Es posible que el modelo sea perjudicial o que las estimaciones se basen en decisiones sesgadas?

En resumen, esta sección busca que los estudiantes apliquen un enfoque crítico para evaluar la interpretabilidad de su modelo, identificar posibles sesgos y analizar las implicaciones de sus decisiones en la población objetivo.



### 7. Concluir [1.0 puntos]

_Aquí deben escribir una breve conclusión del trabajo que hicieron en donde incluyan (pero no se limiten) a responder las siguientes preguntas:_

- ¿Pudieron resolver exitosamente el problema?
- ¿Son aceptables los resultados obtenidos?
- ¿En que medida el EDA ayudó a comprender los datos en miras de generar un modelo predictivo?

Respecto a la clasificación:

- ¿Como fue el rendimiento del baseline para la clasificación?
- ¿Pudieron optimizar el baseline para la clasificación?
- ¿Que tanto mejoro el baseline de la clasificación con respecto a sus optimizaciones?

Finalmente:

- ¿Estuvieron conformes con sus resultados?
- ¿Creen que hayan mejores formas de modelar el problema?
- ¿En general, qué aprendieron del proyecto? ¿Qué no aprendieron y les gustaría haber aprendido?

**OJO** si usted decide responder parte de estas preguntas, debe redactarlas en un formato de informe y no responderlas directamente.

### Otras Instrucciones

Recordar el uso de buenas prácticas de MLOPS como replicabilidad (fijar semillas aleatorias) o el uso del registro de experimentos (con MLFlow). Si bien son opcionales, es altamente recomendado su uso.



### 8. Bonus: Implementación de Kedro y FastAPI [1.5 puntos]

**OPCIONAL**

En esta sección se les solicita utilizar las últimas tecnologías vistas en el curso para la productivización del proyecto de ciencia de datos, centrándose en la organización y gestión de los flujos de trabajo a través de componentes y pipelines, más el servicio del modelo a través del desarrollo de una API.

Para esto: 

1. Genere un proyecto de `Kedro` en donde separe por responsabilidades los nodos/componentes de su proyecto de ciencia de datos en módulos separados. [1.0 puntos]
2. Genere un servidor basado en `FastAPI` el cuál a través de un método post, reciba un batch de datos y genere predicciones para cada uno de ellos. [0.5 puntos]

Las implementaciones son libres. Es decir, usted decide qué componentes implementar, como usar el catálogo de datos y la parametrización del flujo. Sin embargo, evaluaremos buen uso de los framework, modularización y separación de responsabilidades.
